Remove JitGenericHandleCache

davidwrighton opened the pull request:
- It was only used for overflow scenarios from the generic dictionary (which don't happen), virtual resolution scenarios for creating delegates, and a few other rare R2R scenarios
- Replace the virtual resolution scenarios with a cache of the affected data in managed code, and move the helpers to managed
  - Use the GenericCache type which was previously only used on NativeAOT for Generic Virtual method resolution paths.
- Just remove the pointless checks from within the various normal generic lookup paths, and since they no longer do anything interesting in their hot paths except erect a helper method frame and call a worker routine, move those jit helpers to managed as well.

davidwrighton committed:
Remove JitGenericHandleCache
- It was only used for overflow scenarios from the generic dictionary (which don't happen), virtual resolution scenarios for creating delegates, and a few other rare R2R scenarios
- Replace the virtual resolution scenarios with a cache of the affected data in managed code, and move the helpers to managed
- Just remove the pointless checks from within the various normal generic lookup paths

Swap to using GenericCache
- Make FlushCurrentCache public on GenericCache, as we need to clear this when we clean up a LoaderAllocator
- Tweak algorithm for computing out of bound slots on the DictionaryLayout

Update crsttypes

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/VirtualDispatchHelpers.cs:
```diff
@@ -0,0 +1,91 @@
+    private static unsafe IntPtr VirtualFunctionPointerSlowpath(object obj, IntPtr classHandle, IntPtr methodHandle)
+    {
+        IntPtr result = JIT_ResolveVirtualFunctionPointer(ObjectHandleOnStack.Create(ref obj), classHandle, methodHandle);
+        lock(s_virtualFunctionPointerCache)
```
Regular locks can call back into user code through SynchronizationManager. It would be better for this to use Lock with `useTrivialWaits=true`.

davidwrighton commented:
@EgorBot -intel
```csharp
using BenchmarkDotNet.Attributes;
using System.Runtime.CompilerServices;

namespace VirtualDispatchChecks
{

    class _1
    {
        public virtual string Method() => "A";
    }

    class _2 : _1
    {
        public override string Method() => "B";
    }

    class _3 : _1
    {
        public override string Method() => "C";
    }

    class _4 : _1
    {
        public override string Method() => "D";
    }

    class _5 : _1
    {
        public override string Method() => "E";
    }
    class _6 : _1
    {
        public override string Method() => "F";
    }
    class _7 : _1
    {
        public override string Method() => "F";
    }

    class _8 : _1
    {
        public override string Method() => "F";
    }

    class _9 : _1
    {
        public override string Method() => "F";
    }

    class _10 : _1
    {
        public override string Method() => "F";
    }

    public class ResolveVirtualRepeatedly
    {
        static _1 _1 = new _1();
        static _1 _2 = new _2();
        static _1 _3 = new _3();
        static _1 _4 = new _4();
        static _1 _5 = new _5();
        static _1 _6 = new _6();
        static _1 _7 = new _7();
        static _1 _8 = new _8();
        static _1 _9 = new _9();
        static _1 _10 = new _10();
        [MethodImpl(MethodImplOptions.NoInlining)]
        static _1 GetA(int index)
        {
            switch (index % 10)
            {
                case 0:
                    return _10;
                case 1: return _1;
                case 2: return _2;
                case 3: return _3;
                case 4: return _4;
                case 5: return _5;
                case 6: return _6;
                case 7: return _7;
                case 8: return _8;
                case 9: return _9;
            }
            return null!;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static Func<string> GetAction(int index)
        {
            return GetA(index).Method;
        }

        [Benchmark]
        public void Get1Virtual()
        {
            GetAction(0);
        }

        [Benchmark]
        public void Get10Virtual()
        {
            GetAction(0);
            GetAction(1);
            GetAction(2);
            GetAction(3);
            GetAction(4);
            GetAction(5);
            GetAction(6);
            GetAction(7);
            GetAction(8);
            GetAction(9);
        }
    }
}

```

EgorBo commented:
@davidwrighton I have a small limitation in my bot (I plan to fix it), but currently the way it works - it clones `main`, build `corerun`, then it applies PR patch, for that it just does `wget https://github.com/dotnet/runtime/pull/106843.patch` (adds .patch to PR URL basically) which may fail for PRs with large changes/commits (typically, squashing commits via force push helps) -- I'll take a look today to use a better approach to build `diff` (to clone actual branch).

davidwrighton commented:
@EgorBot -intel
```csharp
using BenchmarkDotNet.Attributes;
using System.Runtime.CompilerServices;

namespace VirtualDispatchChecks
{

    class _1
    {
        public virtual string Method() => "A";
    }

    class _2 : _1
    {
        public override string Method() => "B";
    }

    class _3 : _1
    {
        public override string Method() => "C";
    }

    class _4 : _1
    {
        public override string Method() => "D";
    }

    class _5 : _1
    {
        public override string Method() => "E";
    }
    class _6 : _1
    {
        public override string Method() => "F";
    }
    class _7 : _1
    {
        public override string Method() => "F";
    }

    class _8 : _1
    {
        public override string Method() => "F";
    }

    class _9 : _1
    {
        public override string Method() => "F";
    }

    class _10 : _1
    {
        public override string Method() => "F";
    }

    public class ResolveVirtualRepeatedly
    {
        static _1 _1 = new _1();
        static _1 _2 = new _2();
        static _1 _3 = new _3();
        static _1 _4 = new _4();
        static _1 _5 = new _5();
        static _1 _6 = new _6();
        static _1 _7 = new _7();
        static _1 _8 = new _8();
        static _1 _9 = new _9();
        static _1 _10 = new _10();
        [MethodImpl(MethodImplOptions.NoInlining)]
        static _1 GetA(int index)
        {
            switch (index % 10)
            {
                case 0:
                    return _10;
                case 1: return _1;
                case 2: return _2;
                case 3: return _3;
                case 4: return _4;
                case 5: return _5;
                case 6: return _6;
                case 7: return _7;
                case 8: return _8;
                case 9: return _9;
            }
            return null!;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static Func<string> GetAction(int index)
        {
            return GetA(index).Method;
        }

        [Benchmark]
        public void Get1Virtual()
        {
            GetAction(0);
        }

        [Benchmark]
        public void Get10Virtual()
        {
            GetAction(0);
            GetAction(1);
            GetAction(2);
            GetAction(3);
            GetAction(4);
            GetAction(5);
            GetAction(6);
            GetAction(7);
            GetAction(8);
            GetAction(9);
        }
    }
}

```

davidwrighton committed:
Tweak details

davidwrighton commented:
@EgorBot -intel -arm64
```csharp
using BenchmarkDotNet.Attributes;
using System.Runtime.CompilerServices;

namespace VirtualDispatchChecks
{

    class _1
    {
        public virtual string Method() => "A";
    }

    class _2 : _1
    {
        public override string Method() => "B";
    }

    class _3 : _1
    {
        public override string Method() => "C";
    }

    class _4 : _1
    {
        public override string Method() => "D";
    }

    class _5 : _1
    {
        public override string Method() => "E";
    }
    class _6 : _1
    {
        public override string Method() => "F";
    }
    class _7 : _1
    {
        public override string Method() => "F";
    }

    class _8 : _1
    {
        public override string Method() => "F";
    }

    class _9 : _1
    {
        public override string Method() => "F";
    }

    class _10 : _1
    {
        public override string Method() => "F";
    }

    public class ResolveVirtualRepeatedly
    {
        static _1 _1 = new _1();
        static _1 _2 = new _2();
        static _1 _3 = new _3();
        static _1 _4 = new _4();
        static _1 _5 = new _5();
        static _1 _6 = new _6();
        static _1 _7 = new _7();
        static _1 _8 = new _8();
        static _1 _9 = new _9();
        static _1 _10 = new _10();
        [MethodImpl(MethodImplOptions.NoInlining)]
        static _1 GetA(int index)
        {
            switch (index % 10)
            {
                case 0:
                    return _10;
                case 1: return _1;
                case 2: return _2;
                case 3: return _3;
                case 4: return _4;
                case 5: return _5;
                case 6: return _6;
                case 7: return _7;
                case 8: return _8;
                case 9: return _9;
            }
            return null!;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static Func<string> GetAction(int index)
        {
            return GetA(index).Method;
        }

        [Benchmark]
        public void Get1Virtual()
        {
            GetAction(0);
        }

        [Benchmark]
        public void Get10Virtual()
        {
            GetAction(0);
            GetAction(1);
            GetAction(2);
            GetAction(3);
            GetAction(4);
            GetAction(5);
            GetAction(6);
            GetAction(7);
            GetAction(8);
            GetAction(9);
        }
    }
}

```

EgorBo commented:
PS: I have fixed the patch applying so commit squashing is no longer needed. Also, you can use `-perf` to get output of `perf` for a benchmark (only one `[Benchmark]` is expected) before/after.

@EgorBot -arm64 -perf
```csharp
using BenchmarkDotNet.Attributes;
using System.Runtime.CompilerServices;

namespace VirtualDispatchChecks
{

    class _1
    {
        public virtual string Method() => "A";
    }

    class _2 : _1
    {
        public override string Method() => "B";
    }

    class _3 : _1
    {
        public override string Method() => "C";
    }

    class _4 : _1
    {
        public override string Method() => "D";
    }

    class _5 : _1
    {
        public override string Method() => "E";
    }
    class _6 : _1
    {
        public override string Method() => "F";
    }
    class _7 : _1
    {
        public override string Method() => "F";
    }

    class _8 : _1
    {
        public override string Method() => "F";
    }

    class _9 : _1
    {
        public override string Method() => "F";
    }

    class _10 : _1
    {
        public override string Method() => "F";
    }

    public class ResolveVirtualRepeatedly
    {
        static _1 _1 = new _1();
        static _1 _2 = new _2();
        static _1 _3 = new _3();
        static _1 _4 = new _4();
        static _1 _5 = new _5();
        static _1 _6 = new _6();
        static _1 _7 = new _7();
        static _1 _8 = new _8();
        static _1 _9 = new _9();
        static _1 _10 = new _10();
        [MethodImpl(MethodImplOptions.NoInlining)]
        static _1 GetA(int index)
        {
            switch (index % 10)
            {
                case 0:
                    return _10;
                case 1: return _1;
                case 2: return _2;
                case 3: return _3;
                case 4: return _4;
                case 5: return _5;
                case 6: return _6;
                case 7: return _7;
                case 8: return _8;
                case 9: return _9;
            }
            return null!;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static Func<string> GetAction(int index)
        {
            return GetA(index).Method;
        }

        [Benchmark]
        public void Get10Virtual()
        {
            GetAction(0);
            GetAction(1);
            GetAction(2);
            GetAction(3);
            GetAction(4);
            GetAction(5);
            GetAction(6);
            GetAction(7);
            GetAction(8);
            GetAction(9);
        }
    }
}

```

davidwrighton commented:
An analysis of these results indicates that
1. While universally slower with the new code, we are close enough for the performance of this relatively rare scenario.
2. Arm64 showed significantly lower performance once, but that appears to be a case of unlucky hash values in the hash function. Since the has function is not meaningfully changed with this PR, this is considered immaterial.
3. When running in actual use, this should be slightly faster. Benchmark.NET results are gathered with tiered compilation disabled, and the codegen for the function is slightly better when tiered.

EgorBo commented:
> Benchmark.NET results are gathered with tiered compilation disabled, and the codegen for the function is slightly better when tiered.

In the bot it's not turned off, in order to disable it, you can pass `--envvars DOTNET_TieredCompilation:0`. (it's standard BDN CLI args). 
E.g. you can see Tier names in https://gist.github.com/EgorBot/df35cdd0ebee41f43a59d1c44cfa5595 (`OptimizedTier1`)

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/LoaderAllocator.cs:
```diff
@@ -34,6 +34,8 @@ internal sealed partial class LoaderAllocatorScout
             if (m_nativeLoaderAllocator == IntPtr.Zero)
                 return;
 
+            VirtualDispatchHelpers.ClearCache();
```
There may be a subtle race condition:
1. the loader allocator is still in use when `VirtualDispatchHelpers.ClearCache` run
2. the thread is rescheduled right `VirtualDispatchHelpers.ClearCache`
3. the loader allocator becomes unreferenced
4. Destroy succeeds

I think we will end up with stale entries in the hashtable.

I am wondering whether this is the reason for the casting cache flush to be done while the runtime is suspended: https://github.com/dotnet/runtime/pull/106843/files#diff-060f68aac95f40dd380ee5a82dd047222232b25646f908c1ea394438e18a8709R604

davidwrighton committed:
Add a VirtualFuncPtr2 function to cover scenarios which don't need 3 parameters for function pointer resolution
- This will hopefully catch up performance with the previous implementation, and indicate whether such a change is worthwhile

Move cache clearing to native code to address concern from Jan around LoaderAllocator cleanup

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into remove_jitgenericcache

davidwrighton commented:
@EgorBot -intel -arm64
```csharp
using BenchmarkDotNet.Attributes;
using System.Runtime.CompilerServices;

namespace VirtualDispatchChecks
{

    class _1
    {
        public virtual string Method() => "A";
    }

    class _2 : _1
    {
        public override string Method() => "B";
    }

    class _3 : _1
    {
        public override string Method() => "C";
    }

    class _4 : _1
    {
        public override string Method() => "D";
    }

    class _5 : _1
    {
        public override string Method() => "E";
    }
    class _6 : _1
    {
        public override string Method() => "F";
    }
    class _7 : _1
    {
        public override string Method() => "F";
    }

    class _8 : _1
    {
        public override string Method() => "F";
    }

    class _9 : _1
    {
        public override string Method() => "F";
    }

    class _10 : _1
    {
        public override string Method() => "F";
    }

    public class ResolveVirtualRepeatedly
    {
        static _1 _1 = new _1();
        static _1 _2 = new _2();
        static _1 _3 = new _3();
        static _1 _4 = new _4();
        static _1 _5 = new _5();
        static _1 _6 = new _6();
        static _1 _7 = new _7();
        static _1 _8 = new _8();
        static _1 _9 = new _9();
        static _1 _10 = new _10();
        [MethodImpl(MethodImplOptions.NoInlining)]
        static _1 GetA(int index)
        {
            switch (index % 10)
            {
                case 0:
                    return _10;
                case 1: return _1;
                case 2: return _2;
                case 3: return _3;
                case 4: return _4;
                case 5: return _5;
                case 6: return _6;
                case 7: return _7;
                case 8: return _8;
                case 9: return _9;
            }
            return null!;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static Func<string> GetAction(int index)
        {
            return GetA(index).Method;
        }

        [Benchmark]
        public void Get1Virtual()
        {
            GetAction(0);
        }

        [Benchmark]
        public void Get10Virtual()
        {
            GetAction(0);
            GetAction(1);
            GetAction(2);
            GetAction(3);
            GetAction(4);
            GetAction(5);
            GetAction(6);
            GetAction(7);
            GetAction(8);
            GetAction(9);
        }
    }
}

```

davidwrighton committed:
Move generic handle helpers to managed as well

davidwrighton committed:
Remove unnecessary work to have a slightly more efficient implementation for creating delegates to virtual methods. If we need to bring it back, its all isolated to this one commit

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into remove_jitgenericcache

davidwrighton committed:
Add support for nested types in the DEFINE_CLASS_U pathway
Make the ordering of parameters correct for JIT_GenericHandleWorker

davidwrighton committed:
Merge branch 'main' into remove_jitgenericcache

davidwrighton committed:
Fix MUSL build error

davidwrighton committed:
Merge branch 'remove_jitgenericcache' of github.com:davidwrighton/runtime into remove_jitgenericcache

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/VirtualDispatchHelpers.cs:
```diff
@@ -0,0 +1,95 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool Equals(VirtualResolutionData other) =>
+            HashCode == other.HashCode &&
+            MethodTable == other.MethodTable &&
```
HashCode should filter out nearly all mismatches, so the rest of the method can assume that the equality check is most likely going to return true. It may help on some architectures to turn the rest of the method to just one conditional branch, something like:
```
        public bool Equals(VirtualResolutionData other) =>
            HashCode == other.HashCode &&
            (((nint)MethodTable - (nint)other.MethodTable) |
            (ClassHandleTargetMethod - other.ClassHandleTargetMethod) |
            (MethodHandle - other.MethodHandle)) == 0;
```

jkotas commented:
> Benchmark results on Intel

I am not able to reproduce this regression on Dev Box Amd. The PR is ~7% faster compared to baseline on that machine.

davidwrighton commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/VirtualDispatchHelpers.cs:
```diff
@@ -0,0 +1,95 @@
+    private static unsafe IntPtr VirtualFunctionPointerSlowpath(object obj, IntPtr classHandle, IntPtr methodHandle)
+    {
+        IntPtr result = JIT_ResolveVirtualFunctionPointer(ObjectHandleOnStack.Create(ref obj), classHandle, methodHandle);
+        s_virtualFunctionPointerCache.TrySet(new VirtualResolutionData(RuntimeHelpers.GetMethodTable(obj), classHandle, methodHandle), result);
```
NativeAOT follows this rule around not caching results (for interface dispatch, but appears to have the same problem for virtual function resolution)  CoreCLR does not appear to me to have logic in its dispatch behavior which avoids caching. Possibly we intended to build it, but I'm pretty sure we didn't. I can build a test case to validate my expectation if you'd like. The performance of the existing IDynamicInterfaceCastable implementation would likely be highly impacted if we change the rules here.

davidwrighton replied:
The result of this is a useable model, although it's not well defined in behavior. The result is that if any type uses `IDynamicInterfaceCastable` ... the implementor of `IDynamicInterfaceCastable.GetInterfaceImplementaiton` really should provide an idempotent implementation for any case where it will successfully return an implementation, or they may have surprising behavior. However, we DO have the correct cache avoidance at cast time. This means that users of types which use `IDynamicInterfaceCastable` experience the expected behavior.

jkotas replied:
> NativeAOT follows this rule around not caching results (for interface dispatch

We may want to fix NativeAOT to match CoreCLR.

davidwrighton replied:
Agreed. I've opened https://github.com/dotnet/runtime/issues/107999 to track this issue

davidwrighton commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -2331,6 +2087,17 @@ NOINLINE HCIMPL3(CORINFO_MethodPtr, JIT_VirtualFunctionPointer_Framed, Object *
+
     pStaticMD->CheckRestore();
 
     // MDIL: If IL specifies callvirt/ldvirtftn it remains a "virtual" instruction
```
We're supposed to support this sort of adjustment as part of ReadyToRun (since we support converting a non-virtual method into a virtual method); however, I just checked, and we don't actually have testing for the case where we construct a delegate from the target method, and the behavior at runtime is incorrect in this scenario. (We do have tests for actual dispatch conversion into a virtual method.) I intend to update the comment here, but leave the logic in place, and file a bug about the R2R behavior.

davidwrighton replied:
Sigh. Nevermind my previous analysis.

1. Converting from a non-virtual to a virtual in ReadyToRun is not a safe change from the point of view of ReadyToRun. See https://learn.microsoft.com/en-us/dotnet/core/compatibility/library-change-rules for the rules that we actively support, so an affirmative test in C# should not be written.
2. However, we still need to ensure that code doesn't fail in surprising ways with ReadyToRun, so we still need this logic if a customer converts from Virtual to Non-Virtual, which is ALSO a breaking change, but we can hit it if a customer fails to follow all of the rules in the above document. Keeping the existing behavior is done so that failure modes of ReadyToRun generated code are still broadly similar to the failure modes of non ReadyToRun code.

davidwrighton committed:
Code review feedback

jkotas reviewed (approved):
Thanks!
